compiler/mir/mirgen.nim

```diff
@@ -0,0 +1,2105 @@
+  t.skipTypes(abstractInst).kind in {tyPtr, tyRef, tyLent, tyVar, tyOpenArray}
+
+# XXX: copied from ``injectdestructors``. Move somewhere common
+proc isCursor(n: PNode): bool =
+  ## Computes whether the expression `n` names a location that is a cursor
+  case n.kind
+  of nkSym:
+    sfCursor in n.sym.flags
+  of nkDotExpr:
+    isCursor(n[1])
+  of nkCheckedFieldExpr:
+    isCursor(n[0])
+  else:
+    false
+
+func endsInNoReturn(n: PNode): bool =
+  ## Tests and returns whether the simple or compound statement `n` ends in a
+  ## no-return statement
+
+  # TODO: this is a patched version of ``sem.endsInNoReturn`` that also
+  #       considers ``nkPragmaBlock``. Move this procedure somewhere common and
+  #       replace ``sem.endsInNoReturn`` with it
+  const SkipSet = {nkStmtList, nkStmtListExpr, nkPragmaBlock}
+  var it {.cursor.} = n
+  while it.kind in SkipSet and it.len > 0:
+    it = it.lastSon
+
+  result = it.kind in nkLastBlockStmts or
+    (it.kind in nkCallKinds and it[0].kind == nkSym and
+     sfNoReturn in it[0].sym.flags)
+
+func canonicalExpr(n: PNode): PNode =
+  ## Returns the canonical expression, i.e. the expression without leading
+  ## pragma blocks or empty statement lists
+  func skipped(n: PNode): PNode =
+    case n.kind
+    of nkPragmaBlock:
+      result = n.lastSon
+    of nkStmtListExpr:
+      # see if all nodes except the last one are empty. If they are, the
+      # ``nkStmtListExpr`` is skipped
+      for i in 0..<n.len-1:
+        if n[i].kind != nkEmpty:
+          # there are statements in the list -> the list needs to be kept
+          return n
+
+      result = n.lastSon
+    else:
+      result = n
+
+  var n {.cursor.} = n
+  while (let it = skipped(n); it != n):
+    n = it
+
+  result = n
+
+func selectWhenBranch(n: PNode, isNimvm: bool): PNode =
+  assert n.kind == nkWhen
+  if isNimvm: n[0][1]
+  else:       n[1][0]
+
+func selectDefKind(s: PSym): range[mnkDef..mnkDefCursor] =
+  ## Returns the kind of definition to use for the given entity
+  if sfCursor in s.flags: mnkDefCursor
+  else:                   mnkDef
+
+func isSome(x: Destination): bool {.inline.} =
+  x.kind != dkNone
+
+func initDestination(n: PNode, isFirst: bool): Destination =
+  result = Destination(kind: dkGen, node: n, isEmpty: isFirst, own: not isCursor(n))
+
+proc typeOrVoid(g: ModuleGraph, t: PType): PType =
+  ## Returns `t` if it's not 'nil' - the ``void`` type otherwise
+  if t != nil: t
+  else:        g.getSysType(unknownLineInfo, tyVoid)
+
+proc typeOrVoid(c: TCtx, t: PType): PType {.inline.} =
+  ## Returns `t` if it's not 'nil' - the ``void`` type otherwise
+  # TODO: cache the void type
+  typeOrVoid(c.graph, t)
+
+func tempNode(typ: PType, id: TempId): MirNode {.inline.} =
+  MirNode(kind: mnkTemp, typ: typ, temp: id)
+
+
+func nextTempId(c: var TCtx): TempId =
+  result = TempId(c.numTemps)
+  inc c.numTemps
+
+func nextLabel(c: var TCtx): LabelId =
+  result = LabelId(c.numLabels + 1)
```

Comment:
I'd recommend `high(LabelId)` as the invalid id if you need those instead of shifting.

Reply:
But then the zero-representation is not the 'none'/invalid value, which is a property that I dislike and that can cause issues easily.